### PR TITLE
Add possibility to configure data network for contrail control

### PR DESCRIPTION
**Pull request by michalskalski**
dataSubnet property describes dedicated subnet used for
establishing xmpp session between contrail control and
external vrouters. If dataSubnet is set:

  * contrail StatefulSets are created with the pod
    template which will creates pods with dataSubnet
    annotations

  * when the init container starts it checks if the
    pod has annotation dataSubnet, if yes it tries
    to discover local pod ip from this subnet and
    store it in pod annotations as dataSubnetIP

  * when dataSubnetIP is discovered in pod annotations
    it will extend subject of the pod's ssl certificate

  * dataSubnetIP will be used for hostip property in
    contrail services configurations - services will
    listen on this IP, and provision script will
    register controllers with their dataSubnetIPs

### CODE REVIEW

**Comment by michalskalski** [comment]
recheck ha

**Review by Dysproz — APPROVED** [review]

File: pkg/certificates/certificate_subject.go [hunk 1]
```diff
@@ -58,6 +63,9 @@ func (c certificateSubject) generateCertificateTemplate(alternativeIP string) (x
 	if alternativeIP != "" {
 		ips = append(ips, net.ParseIP(alternativeIP))
 	}
+	for _, ip := range c.secondaryIPs {
```
Comment (katrybacka) [inline pkg/certificates/certificate_subject.go:66]: I guess that alternativeIP (service IP indeed) is pretty the same as this dataSubnetIP, so it the certification generation mechanism can be unified and this alternativeIP could be a part of secondaryIPs

File: pkg/controller/control/control_controller_test.go [hunk 2]
```diff
@@ -131,30 +123,43 @@ func TestControlController(t *testing.T) {
...
+			Namespace: controlName.Namespace,
+		}, sts)
+		assert.NoError(t, err)
+		expectedAnnotation := map[string]string{"dataSubnet": "172.17.90.0/24"}
+		assert.Truef(
+			t,
+			reflect.DeepEqual(sts.Spec.Template.Annotations, expectedAnnotation),
+			"expected annotation not applied, expected: %v, got: %v",
+			expectedAnnotation,
+			sts.Spec.Template.Annotations,
```
Comment (katrybacka) [inline pkg/controller/control/control_controller_test.go:162]: assert.Equal is not enough for this check?

**Review by katrybacka — CHANGES_REQUESTED** [review]

File: pkg/apis/contrail/v1alpha1/control_types.go [hunk 3]
```diff
@@ -263,41 +274,45 @@ func (c *Control) InstanceConfiguration(request reconcile.Request,
...
 			ListenAddress string
+			DataIP        string
 			APIServerList string
 			ASNNumber     string
 			BGPPort       string
 			APIServerPort string
 			Hostname      string
 		}{
-			ListenAddress: podList.Items[idx].Status.PodIP,
+			ListenAddress: podIP,
```
Comment (kaweue) [inline pkg/apis/contrail/v1alpha1/control_types.go:307]: Is now "ListenAddress" field used?
Why not just use this ListenAddress and assign dataIP there.
`ListenAddress: dataIP`

Technically it is listen address of this service am I right?

File: pkg/apis/contrail/v1alpha1/control_types.go [hunk 4]
```diff
@@ -263,41 +274,45 @@ func (c *Control) InstanceConfiguration(request reconcile.Request,
 			RabbitmqVhost:       rabbitmqSecretVhost,
 			CAFilePath:          certificates.SignerCAFilepath,
 		})
-		data["dns."+podList.Items[idx].Status.PodIP] = controlDNSConfigBuffer.String()
+		data["dns."+podIP] = controlDNSConfigBuffer.String()
 
 		var controlNodemanagerBuffer bytes.Buffer
 		configtemplates.ControlNodemanagerConfig.Execute(&controlNodemanagerBuffer, struct {
 			ListenAddress       string
+			DataIP              string
```
Comment (kaweue) [inline pkg/apis/contrail/v1alpha1/control_types.go:282]: Is this value used in this "ControlNodemanagerConfig" template?

**Review by kaweue — CHANGES_REQUESTED** [review]

File: pkg/certificates/certificate_subject.go [hunk 5]
```diff
@@ -32,7 +33,11 @@ func (cs certificateSubjects) createCertificateSubjects() []certificateSubject {
 		} else {
 			hostname = pod.Spec.Hostname
 		}
-		subjects = append(subjects, certificateSubject{name: pod.Name, hostname: hostname, ip: pod.Status.PodIP})
+		if dataIP, isSet := pod.Annotations["dataSubnetIP"]; isSet {
```
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/certificates/certificate_subject.go:36]: I agree with you, certificates logic shouldn't know about internals of control pods. In first approach I tried do this in the way of passing those IPs as separate field in certificate type but those IP are slightly different than ServiceIP, which is a common for a group of Pods. DataSubnetIP is specific for a pod, and we iterate over them here. If we just pass list of those IPs we could lose connection between IP and pod to which it belongs. I could maybe create a map which would show pod to IP connection and pass it here, which maybe is more general, but recreates the connection which already exists. I could name this annotation to be more general like the hostname is, and assume that all Pods could have it, but probably that will not be use outside of control pods.

File: pkg/certificates/certificate_subject.go (hunk 1 above)
Comment (michalskalski) [reply pkg/certificates/certificate_subject.go:66]: The difference between them is that serviceIP is common for all Pods of given type, and dataSubnetIP is specific for the Pod. I had a problem to maintain connection between pod and it's IP which prevented me from merging those two information in one. I also discussing about it with @kaweue here https://github.com/Juniper/contrail-operator/pull/398#discussion_r494900683

File: pkg/apis/contrail/v1alpha1/templates/control_config.go [hunk 6]
```diff
@@ -9,7 +9,7 @@ bgp_port=179
 collectors={{ .CollectorServerList }}
 # gr_helper_bgp_disable=0
 # gr_helper_xmpp_disable=0
-hostip={{ .ListenAddress }}
+hostip={{ .DataIP }}
```
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/apis/contrail/v1alpha1/templates/control_config.go:12]: Also thought about this. When deployed with "0.0.0.0" it was a first address which was displayed in webui, in details you could see the IP used for provisioning. I'm also not sure if it has impact on establishing bgp session, how is controller presents to the peers. The way I chose to implement it was the replication of how it is configured in Contrail Cloud, the approach with "0.0.0.0" seemed to work but I don't have that much experience it this area to confirm it is proper solution.

[the remark this answers is not in the document]
Comment (kaweue) [reply pkg/apis/contrail/v1alpha1/templates/control_config.go:12]: Ok, understood. That it is not only listen address but also it used for something else.

File: pkg/certificates/certificate_subject.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaweue) [reply pkg/certificates/certificate_subject.go:36]: Ok, more I look at this I think that there is problem with api of certificates package. There are methods like `NewCertificateWithServiceIP` and `NewCertificate` which accepts the set of pods and in the end they are creating a set of certificates of those pods. I'm not sure if that is really good. Maybe if the API could allow to user to create individual certificates for individual pods we could extend it easily.

Maybe API like this
```## Loop for every POD
  certificates.NewCertificates(r.client, r.scheme, command, instanceType).Cert(hostname, ip, alterantiveIPs).EnsureExits()

File: pkg/apis/contrail/v1alpha1/control_types.go [hunk 7]
```diff
@@ -55,6 +55,11 @@ type ControlConfiguration struct {
 	RabbitmqUser      string       `json:"rabbitmqUser,omitempty"`
 	RabbitmqPassword  string       `json:"rabbitmqPassword,omitempty"`
 	RabbitmqVhost     string       `json:"rabbitmqVhost,omitempty"`
+	// DataSubnet allow to set alternative network in which control, nodemanager
+	// and dns services will listen. Local pod address from this subnet will be
+	// discovered and used both in configuration for hostip directive and provision
+	// script.
+	DataSubnet string `json:"dataSubnet,omitempty"`
```
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/apis/contrail/v1alpha1/control_types.go:63]: good idea, added

File: pkg/apis/contrail/v1alpha1/control_types.go [hunk 8]
```diff
@@ -183,13 +188,15 @@ func (c *Control) InstanceConfiguration(request reconcile.Request,
 	var data = make(map[string]string)
 	for idx := range podList.Items {
 		hostname := podList.Items[idx].Annotations["hostname"]
+		dataIP := getDataIP(&podList.Items[idx])
```
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/apis/contrail/v1alpha1/control_types.go]: fixed, yes now it is cleaner

File: pkg/controller/control/control_controller_test.go (hunk 2 above)
Comment (michalskalski) [reply pkg/controller/control/control_controller_test.go]: it is enough, it already use DeepEqual underneath, thank you for pointing that out

File: pkg/apis/contrail/v1alpha1/base_types.go [hunk 9]
```diff
@@ -622,7 +623,10 @@ func getPodInitStatus(reconcileClient client.Client,
 				if initStatus.Name == "init" {
 					if initStatus.State.Terminated == nil {
 						if initStatus.State.Running != nil {
-							var annotationMap = make(map[string]string)
+							annotationMap := podList.Items[idx].GetAnnotations()
```
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/apis/contrail/v1alpha1/base_types.go]: fixed

File: pkg/certificates/certificate_subject.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/certificates/certificate_subject.go:36]: ok I will look into it

File: pkg/apis/contrail/v1alpha1/templates/control_config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalskalski) [reply pkg/apis/contrail/v1alpha1/templates/control_config.go]: Listen on 0.0.0.0 does not seem to have impact on bgp/xmpp session establish process. With last commit I switched to this value in hostip. Code is maybe more clear now.

**Comment by michalskalski** [comment]
recheck

**Review by kaweue — APPROVED** [review]

**Comment by michalskalski** [comment]
@katrybacka I think I addressed your comments, is it now looks ok for you?

**Comment by katrybacka** [comment]
@michalskalski I checked only my comments and it looks good, I had not time for a full review but this part is ok :+1:

**Review by katrybacka — APPROVED** [review]

**Pull request merged**